extensions/git/src/gitEditor.ts

```diff
@@ -0,0 +1,75 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as path from 'path';
+import { workspace, window, Uri, Position, Selection } from 'vscode';
+
+import { IIPCHandler, IIPCServer } from './ipc/ipcServer';
+import { IDisposable } from './util';
+
+export interface GitEditorEnvironment {
+	GIT_EDITOR: string;
+	ELECTRON_RUN_AS_NODE?: string;
+	VSCODE_GIT_EDITOR_NODE?: string;
+	VSCODE_GIT_EDITOR_MAIN?: string;
+}
+
+interface GitEditorRequest {
+	commitMessagePath?: string;
+}
+
+export class GitEditor implements IIPCHandler {
+	private disposable: IDisposable;
+
+	static getDisabledEnv(): GitEditorEnvironment {
+		const fileType = process.platform === 'win32' ? 'bat' : 'sh';
+		const gitEditor = path.join(__dirname, `scripts/git-editor-empty.${fileType}`);
+
+		return {
+			GIT_EDITOR: `'${gitEditor}'`,
+		};
+	}
+
+	constructor(ipc: IIPCServer) {
+		this.disposable = ipc.registerHandler('git-editor', this);
+	}
+
+	async handle({ commitMessagePath }: GitEditorRequest): Promise<any> {
+		if (commitMessagePath) {
+			const file = Uri.file(commitMessagePath);
+			const doc = await workspace.openTextDocument(file);
+			const editor = await window.showTextDocument(doc);
+
+			// We don't want to remember the cursor position.
+			// One should be able to start writing the message immediately.
+			const position = new Position(0, 0);
+			editor.selection = new Selection(position, position);
```

Comment:
vscode seems to remember where my cursor was in an older file. This was quite annoying since I couldn't start writing the commit message immediately.

Reply:
This is probably because the path to the git-commit is always `.git/COMMIT_EDITMSG` or something. If you change how vscode calls git commit, you can make this a randomly named temp file. The downside is that you will lose the automated report that git commit gives you (especially when using `--verbose`) but then again, you have a whole IDE that can provide this information so maybe that's a good future path.

Reply:
> This is probably because the path to the git-commit is always .git/COMMIT_EDITMSG or something. If you change how vscode calls git commit, you can make this a randomly named temp file. 

does it mean that not using .git/COMMIT_EDITMSG means no proper integration with prepare-commit-msg hook? it would be nice to preserve this

Reply:
Yes, that's why I refrained from using a custom/tmp file. I think the overall experience is still the best when the original file is used.